templates/profile/apache/tnsnames.ora.erb

```diff
@@ -0,0 +1,48 @@
+<% @oracle_servers.each do | server, server_alias_arr | -%>
+# <%= server %>
+#
+<% server_alias_arr.each do | server_alias | -%>
+<%= server_alias.upcase %> =
+  (DESCRIPTION =
+    (ADDRESS =
+      (PROTOCOL = TCP)
+      (HOST = <%= server.downcase %>.umdl.umich.edu)
```

Comment:
Is there a reason to enforce upcase and downcase here? For the principle of least surprising behavior, I'd recommend removing all `.upcase` and `.downcase` calls in favor of upcasing/downcasing the hieradata, so that what's in the hieradata will be copied faithfully into the configuration.

It'd be different if we were sometimes running `server.upcase` or `server_alias.downcase`, but I don't see that anywhere.

Reply:
I agree with your use case for certain. That is the main use case and I did think about that in this context.

So server_alias needs to be uppercase and the server name lowercase at least by config file convention if not technical requirement.

What prompted the upcase and downcase was to compensate for human error and cognitive overload of having to remember that we need the server name to be lowercase but the server aliases to be uppercase. With this the code handles that consistency checking and fixing so we don't have to when adding node files.

Reply:
Thoughts given that info?

Basically if I don't do the code then it will need us to remember to write stuff like:

myserver:
  - SERVERALIAS

And not any variation from that.

Reply:
Thanks for the explanation. That makes sense, although I still think it's out of bounds for what puppet is for. Since we're talking about generating a configuration file, we should understand what goes in that file when we're writing configuration regardless of whether that configuration lives in a file on a server or in a yaml file.

I guess what I'm saying is that I think our hiera configuration has more value if we keep it faithful to what it comes out as. If we start adding modifiers, that adds harder-to-predict behavior that I think is worth it when we're removing configuration from hieradata or at least generalizing it.

If you're worried about lowercase letters in the configuration causing breaking changes, a better approach could be to add something that causes the profile to fail with an error instead of generating the file at all, like this:

```puppet
if <any of the aliases contain lowercase letters> {
  fail('Oracle server aliases must be all uppercase')
}
```

This way, we require our configuration to remain faithful to reality.

Reply:
I agree. It wasn't fully settling with me either. 

I like your suggested fix better. :) I will look into this once I finish troubleshooting why it isn't working currently in testing.

Reply:
If you want, you could also add a test with `it { is_expected.not_to compile }`

Sometimes it feels pretty good to have a "this shouldn't compile" test and watch it pass.